File: sdk/python/src/roche_sandbox/intent.py
# SPDX-License-Identifier: Apache-2.0
# Copyright 2025 Substratum Labs

"""Client-side intent analysis for code execution.

Python uses AST parsing for precise import/call analysis.
Node and Bash use keyword matching.
"""

from __future__ import annotations

import ast
import re
from dataclasses import dataclass, field


@dataclass
class CodeIntent:
    """Result of analyzing code intent."""

    provider: str = "docker"
    """Recommended provider: 'wasm', 'docker', 'firecracker'."""
    needs_network: bool = False
    network_hosts: list[str] = field(default_factory=list)
    needs_writable: bool = False
    writable_paths: list[str] = field(default_factory=list)
    needs_packages: bool = False
    package_manager: str | None = None
    memory_hint: str | None = None
    language: str = "python"
    confidence: float = 0.5
    reasoning: list[str] = field(default_factory=list)


# URL pattern for host extraction
_URL_RE = re.compile(r"https?://([a-zA-Z0-9.-]+)")

# --- Python AST-based analysis ---

# Modules that imply network access
_PY_NETWORK_MODULES = {
    "requests", "urllib", "urllib.request", "urllib3", "httpx",
    "aiohttp", "socket", "http", "http.client", "http.server",
    "ftplib", "smtplib", "xmlrpc", "grpc",
}

# Modules that imply package installation (heavy deps unlikely pre-installed)
_PY_HEAVY_PACKAGES = {
    "pandas", "numpy", "scipy", "sklearn", "scikit-learn",
    "tensorflow", "torch", "pytorch", "matplotlib", "seaborn",
    "polars", "xgboost", "lightgbm", "transformers", "langchain",
    "openai", "anthropic",
}

# Modules that imply high memory
_PY_HEAVY_MEMORY = {
    "pandas", "numpy", "scipy", "tensorflow", "torch", "polars",
    "xgboost", "lightgbm", "transformers",
}

# Functions/methods that imply filesystem writes
_PY_WRITE_FUNCS = {
    "open",  # checked with write mode
}

_PY_WRITE_METHODS = {
    "write", "writelines", "to_csv", "to_json", "to_parquet",
    "to_excel", "to_pickle", "to_hdf", "save", "savefig",
    "dump",  # json.dump, pickle.dump
}

_PY_WRITE_CALLS = {
    "os.makedirs", "os.mkdir", "os.rename", "os.remove", "os.unlink",
    "shutil.copy", "shutil.move", "shutil.rmtree",
    "pathlib.Path.mkdir", "pathlib.Path.write_text", "pathlib.Path.write_bytes",
}

# Modules that imply subprocess/shell execution
_PY_SUBPROCESS_MODULES = {"subprocess", "os"}

_PY_SUBPROCESS_FUNCS = {
    "subprocess.run", "subprocess.call", "subprocess.Popen",
    "subprocess.check_output", "subprocess.check_call",
    "os.system", "os.popen", "os.exec", "os.execvp",
}


def _analyze_python_ast(intent: CodeIntent, code: str) -> bool:
    """Analyze Python code via AST. Returns True if successful."""
    try:
        tree = ast.parse(code)
    except SyntaxError:
        return False

    imports = _extract_imports(tree)
    calls = _extract_calls(tree)
    strings = _extract_strings(tree)

    # --- Imports → network, packages, memory ---
    for mod in imports:
        top = mod.split(".")[0]

        if mod in _PY_NETWORK_MODULES or top in _PY_NETWORK_MODULES:
            intent.needs_network = True
            intent.reasoning.append(f"Network: import `{mod}`")

        if top in _PY_HEAVY_PACKAGES:
            intent.needs_packages = True
            intent.needs_network = True
            intent.package_manager = "pip"
            intent.reasoning.append(f"Package: import `{mod}`")
            if "pypi.org" not in intent.network_hosts:
                intent.network_hosts.append("pypi.org")

        if top in _PY_HEAVY_MEMORY and intent.memory_hint is None:
            intent.memory_hint = "512m"
            intent.reasoning.append(f"Memory 512m: `{top}`")

    # --- Calls → filesystem writes, subprocess ---
    for call_name, args in calls:
        # open() with write mode
        if call_name == "open" and len(args) >= 2:
            mode = args[1]
            if isinstance(mode, str) and any(c in mode for c in "wax+"):
                intent.needs_writable = True
                intent.reasoning.append(f"FS write: open({args[0]!r}, {mode!r})")
                if isinstance(args[0], str):
                    _add_writable_path(intent, args[0])

        # .write(), .to_csv(), etc.
        parts = call_name.rsplit(".", 1)
        if len(parts) == 2 and parts[1] in _PY_WRITE_METHODS:
            intent.needs_writable = True
            intent.reasoning.append(f"FS write: .{parts[1]}()")

        # os.makedirs, shutil.copy, etc.
        if call_name in _PY_WRITE_CALLS:
            intent.needs_writable = True
            intent.reasoning.append(f"FS write: {call_name}()")

        # subprocess → likely needs network or writable
        if call_name in _PY_SUBPROCESS_FUNCS:
            intent.reasoning.append(f"Subprocess: {call_name}()")
            # Check if subprocess runs pip/apt
            if args and isinstance(args[0], (str, list)):
                cmd_str = str(args[0])
                if "pip install" in cmd_str or "pip3 install" in cmd_str:
                    intent.needs_packages = True
                    intent.needs_network = True
                    intent.package_manager = "pip"
                    intent.reasoning.append("Package: subprocess pip install")
                if any(x in cmd_str for x in ["apt-get install", "apt install"]):
                    intent.needs_packages = True
                    intent.needs_network = True
                    intent.package_manager = "apt"

    # --- Strings → URLs/hosts ---
    for s in strings:
        for match in _URL_RE.finditer(s):
            host = match.group(1)
            if host not in intent.network_hosts:
                intent.network_hosts.append(host)
                intent.needs_network = True
                intent.reasoning.append(f"Host: {host}")

    # --- String paths → writable paths ---
    if intent.needs_writable:
        for s in strings:
            _add_writable_path(intent, s)

    if intent.needs_writable and not intent.writable_paths:
        intent.writable_paths.append("/tmp")

    intent.confidence = 0.9  # AST is high confidence
    return True


def _extract_imports(tree: ast.AST) -> set[str]:
    """Extract all imported module names from AST."""
    modules: set[str] = set()
    for node in ast.walk(tree):
        if isinstance(node, ast.Import):
            for alias in node.names:
                modules.add(alias.name)
        elif isinstance(node, ast.ImportFrom):
            if node.module:
                modules.add(node.module)
    return modules


def _extract_calls(tree: ast.AST) -> list[tuple[str, list]]:
    """Extract function calls with their string/list arguments."""
    calls: list[tuple[str, list]] = []
    for node in ast.walk(tree):
        if not isinstance(node, ast.Call):
            continue
        name = _call_name(node)
        if name is None:
            continue
        # Extract constant args for analysis
        args = []
        for arg in node.args:
            if isinstance(arg, ast.Constant) and isinstance(arg.value, (str, int, float)):
                args.append(arg.value)
            elif isinstance(arg, ast.List):
                elts = [e.value for e in arg.elts if isinstance(e, ast.Constant)]
                args.append(elts)
            else:
                args.append(None)
        calls.append((name, args))
    return calls


def _call_name(node: ast.Call) -> str | None:
    """Extract dotted call name: 'os.makedirs', 'open', 'df.to_csv'."""
    func = node.func
    if isinstance(func, ast.Name):
        return func.id
    if isinstance(func, ast.Attribute):
        parts = [func.attr]
        obj = func.value
        while isinstance(obj, ast.Attribute):
            parts.append(obj.attr)
            obj = obj.value
        if isinstance(obj, ast.Name):
            parts.append(obj.id)
        parts.reverse()
        return ".".join(parts)
    return None


def _extract_strings(tree: ast.AST) -> list[str]:
    """Extract all string constants from AST."""
    strings: list[str] = []
    for node in ast.walk(tree):
        if isinstance(node, ast.Constant) and isinstance(node.value, str) and len(node.value) > 3:
            strings.append(node.value)
    return strings


def _add_writable_path(intent: CodeIntent, s: str) -> None:
    """Add writable path if string looks like a path."""
    for prefix in ["/tmp", "/output", "/data", "/workspace", "/app", "/home"]:
        if s.startswith(prefix) and prefix not in intent.writable_paths:
            intent.writable_paths.append(prefix)


# --- Non-Python: keyword-based analysis ---

_KW_NETWORK: dict[str, list[str]] = {
    "python": ["import requests", "import urllib", "import httpx", "import aiohttp",
               "from requests", "from urllib", "from httpx", "from aiohttp",
               "import socket", "import http.client"],
    "node": ["require('http')", "require('https')", "require('axios')",
             "require('node-fetch')", "import fetch", "import axios"],
    "bash": ["curl ", "wget ", "nc ", "ssh "],
}
_KW_WRITE: dict[str, list[str]] = {
    "python": ["open(", "with open", ".write(", ".to_csv(", ".to_json(",
               "os.makedirs(", "os.mkdir("],
    "node": ["fs.writeFile", "fs.writeSync", "fs.mkdir", "createWriteStream"],
    "bash": [" > ", " >> ", "mkdir ", "touch ", "tee "],
}
_KW_HEAVY_LIBS = ["pandas", "numpy", "scipy", "tensorflow", "torch", "polars"]

_PACKAGE_INDICATORS: list[tuple[str, str, list[str]]] = [
    ("python", "pip", ["pip install", "pip3 install"]),
    ("python", "pip", ["import pandas", "import numpy", "import scipy",
                       "import sklearn", "import tensorflow", "import torch"]),
    ("node", "npm", ["npm install", "npx "]),
    ("bash", "apt", ["apt-get install", "apt install", "yum install", "apk add"]),
]

# Package registries
_REGISTRIES = {"pip": "pypi.org", "npm": "registry.npmjs.org"}


def _analyze_keyword(intent: CodeIntent, code: str, language: str) -> None:
    """Keyword-based analysis (fallback for Python, primary for Node/Bash)."""
    # Network
    for lang, indicators in _KW_NETWORK.items():
        if lang != language and language != "auto":
            continue
        for ind in indicators:
            if ind in code:
                intent.needs_network = True
                intent.reasoning.append(f"Network: `{ind}`")
                break

    # Filesystem writes
    for lang, indicators in _KW_WRITE.items():
        if lang != language and language != "auto":
            continue
        for ind in indicators:
            if ind in code:
                intent.needs_writable = True
                intent.reasoning.append(f"FS write: `{ind}`")
                break

    # Memory hints
    for lib in _KW_HEAVY_LIBS:
        if lib in code and intent.memory_hint is None:
            intent.memory_hint = "512m"
            intent.reasoning.append(f"Memory 512m: `{lib}`")
            break

    # Package install
    for lang, pm, inds in _PACKAGE_INDICATORS:
        if lang != language and language != "auto":
            continue
        for ind in inds:
            if ind in code:
                intent.needs_packages = True
                intent.needs_network = True
                intent.package_manager = pm
                intent.reasoning.append(f"Package: `{ind}`")
                registry = _REGISTRIES.get(pm, "")
                if registry and registry not in intent.network_hosts:
                    intent.network_hosts.append(registry)
                break

    # URLs
    for match in _URL_RE.finditer(code):
        host = match.group(1)
        if host not in intent.network_hosts:
            intent.network_hosts.append(host)
            intent.needs_network = True
            intent.reasoning.append(f"Host: {host}")

    # Writable paths
    if intent.needs_writable:
        for path in ["/tmp", "/output", "/data", "/workspace"]:
            if path in code and path not in intent.writable_paths:
                intent.writable_paths.append(path)
        if not intent.writable_paths:
            intent.writable_paths.append("/tmp")


# --- Main entry point ---


def analyze(code: str, language: str = "python") -> CodeIntent:
    """Analyze code and infer execution intent.

    Python: uses AST parsing (precise, no false positives from comments).
    Node/Bash: uses keyword matching.
    """
    intent = CodeIntent(language=language, provider="wasm")

    if language == "python" or language == "auto":
        if _analyze_python_ast(intent, code):
            # AST succeeded — also do URL extraction from non-Python if auto
            if language == "auto":
                _analyze_keyword(intent, code, "auto")
        else:
            # AST failed (syntax error) — fall back to keywords
            _analyze_keyword(intent, code, "python")
            intent.reasoning.append("AST parse failed, fell back to keyword matching")
    else:
        _analyze_keyword(intent, code, language)

    _determine_provider(intent)
    return intent


def _determine_provider(intent: CodeIntent) -> None:
    if intent.needs_network or intent.needs_packages or intent.needs_writable:
        intent.provider = "docker"
        if intent.confidence < 0.8:
            intent.confidence = 0.8
    else:
        intent.provider = "wasm"
        if intent.confidence < 0.7:
            intent.confidence = 0.7
    intent.reasoning.append(f"Provider: {intent.provider} ({intent.confidence:.0%})")

File: crates/roche-core/src/intent.rs
// SPDX-License-Identifier: Apache-2.0
// Copyright 2025 Substratum Labs

//! Intent-based code analysis engine.
//!
//! Analyzes code to infer:
//! - Best provider (WASM for pure compute, Docker for packages/network, Firecracker for isolation)
//! - Minimal permission set (network allowlist, writable paths)
//! - Resource hints (memory, timeout)

use serde::{Deserialize, Serialize};

/// Result of analyzing code intent.
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct CodeIntent {
    /// Recommended provider.
    pub provider: ProviderHint,
    /// Whether network access is needed.
    pub needs_network: bool,
    /// Specific hosts the code will likely contact.
    pub network_hosts: Vec<String>,
    /// Whether filesystem writes are needed.
    pub needs_writable: bool,
    /// Specific paths that need write access.
    pub writable_paths: Vec<String>,
    /// Whether package installation is detected.
    pub needs_packages: bool,
    /// Detected package manager.
    pub package_manager: Option<String>,
    /// Suggested memory hint (e.g., "256m" for data-heavy code).
    pub memory_hint: Option<String>,
    /// Detected language.
    pub language: String,
    /// Confidence score 0.0–1.0.
    pub confidence: f64,
    /// Human-readable explanation of the analysis.
    pub reasoning: Vec<String>,
}

/// Provider recommendation.
#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
pub enum ProviderHint {
    /// Pure compute, no I/O — can use WASM for sub-ms startup.
    Wasm,
    /// Needs packages, network, or system tools — use Docker.
    #[default]
    Docker,
    /// Untrusted code needing strong isolation — use Firecracker.
    Firecracker,
}

/// Analyze code and infer execution intent.
pub fn analyze(code: &str, language: &str) -> CodeIntent {
    let mut intent = CodeIntent {
        language: language.to_string(),
        provider: ProviderHint::Wasm, // start optimistic
        confidence: 0.5,
        ..Default::default()
    };

    // Run all analyzers
    analyze_network(&mut intent, code, language);
    analyze_packages(&mut intent, code, language);
    analyze_filesystem(&mut intent, code, language);
    analyze_resources(&mut intent, code, language);

    // Determine provider based on what we found
    determine_provider(&mut intent);

    intent
}

fn analyze_network(intent: &mut CodeIntent, code: &str, language: &str) {
    // Detect HTTP/network libraries
    let network_indicators: &[(&str, &[&str])] = &[
        (
            "python",
            &[
                "import requests",
                "import urllib",
                "import httpx",
                "import aiohttp",
                "from requests",
                "from urllib",
                "from httpx",
                "from aiohttp",
                "import socket",
                "import http.client",
                "import http.server",
                "import urllib3",
                "import ftplib",
                "import smtplib",
                "import xmlrpc",
                "import grpc",
                "from urllib3",
                "from grpc",
            ],
        ),
        (
            "node",
            &[
                "require('http')",
                "require('https')",
                "require('axios')",
                "require('node-fetch')",
                "import fetch",
                "import axios",
            ],
        ),
        ("bash", &["curl ", "wget ", "nc ", "ssh "]),
    ];

    for (lang, indicators) in network_indicators {
        if *lang == language || language == "auto" {
            for indicator in *indicators {
                if code.contains(indicator) {
                    intent.needs_network = true;
                    intent
                        .reasoning
                        .push(format!("Network access needed: found `{indicator}`"));
                    break;
                }
            }
        }
    }

    // Extract URLs/hosts from code
    extract_hosts(intent, code);
}

fn extract_hosts(intent: &mut CodeIntent, code: &str) {
    // Simple pattern: look for quoted strings containing domain-like patterns
    let patterns = ["https://", "http://", "api.", "www."];

    for line in code.lines() {
        for pattern in &patterns {
            if let Some(pos) = line.find(pattern) {
                if let Some(host) = extract_host_from_url(&line[pos..]) {
                    if !intent.network_hosts.contains(&host) {
                        intent.network_hosts.push(host.clone());
                        intent.needs_network = true;
                        intent.reasoning.push(format!("Detected host: {host}"));
                    }
                }
            }
        }
    }
}

fn extract_host_from_url(s: &str) -> Option<String> {
    // Strip protocol
    let s = s
        .strip_prefix("https://")
        .or_else(|| s.strip_prefix("http://"))
        .unwrap_or(s);

    // Take until first non-hostname char
    let host: String = s
        .chars()
        .take_while(|c| c.is_alphanumeric() || *c == '.' || *c == '-' || *c == ':')
        .collect();

    // Strip port
    let host = host.split(':').next().unwrap_or(&host);

    // Basic validation: must have a dot and be reasonable length
    if host.contains('.') && host.len() > 3 && host.len() < 256 {
        Some(host.to_string())
    } else {
        None
    }
}

fn analyze_packages(intent: &mut CodeIntent, code: &str, language: &str) {
    let package_indicators: &[(&str, &str, &[&str])] = &[
        (
            "python",
            "pip",
            &["pip install", "pip3 install", "import subprocess"],
        ),
        (
            "python",
            "pip",
            &[
                "import pandas",
                "import numpy",
                "import scipy",
                "import sklearn",
                "import tensorflow",
                "import torch",
                "import matplotlib",
                "import seaborn",
                "import polars",
                "import xgboost",
                "import lightgbm",
                "import transformers",
                "import langchain",
                "import openai",
                "import anthropic",
            ],
        ),
        ("node", "npm", &["npm install", "npx ", "require('"]),
        (
            "bash",
            "apt",
            &["apt-get install", "apt install", "yum install", "apk add"],
        ),
    ];

    for (lang, pm, indicators) in package_indicators {
        if *lang == language || language == "auto" {
            for indicator in *indicators {
                if code.contains(indicator) {
                    intent.needs_packages = true;
                    intent.needs_network = true;
                    intent.package_manager = Some(pm.to_string());
                    intent
                        .reasoning
                        .push(format!("Package install needed: found `{indicator}`"));

                    // Add package registry to hosts
                    let registry = match *pm {
                        "pip" => "pypi.org",
                        "npm" => "registry.npmjs.org",
                        _ => "",
                    };
                    if !registry.is_empty() && !intent.network_hosts.contains(&registry.to_string())
                    {
                        intent.network_hosts.push(registry.to_string());
                    }
                    break;
                }
            }
        }
    }
}

fn analyze_filesystem(intent: &mut CodeIntent, code: &str, language: &str) {
    let write_indicators: &[(&str, &[&str])] = &[
        (
            "python",
            &[
                "open(",
                "with open",
                ".to_csv(",
                ".to_json(",
                ".to_parquet(",
                ".to_excel(",
                ".to_pickle(",
                ".savefig(",
                "os.makedirs(",
                "os.mkdir(",
                "os.rename(",
                "os.remove(",
                "shutil.copy(",
                "shutil.move(",
                "shutil.rmtree(",
                "pathlib",
            ],
        ),
        (
            "node",
            &[
                "fs.writeFile",
                "fs.writeSync",
                "fs.mkdir",
                "createWriteStream",
            ],
        ),
        ("bash", &[" > ", " >> ", "mkdir ", "touch ", "tee "]),
    ];

    for (lang, indicators) in write_indicators {
        if *lang == language || language == "auto" {
            for indicator in *indicators {
                if code.contains(indicator) {
                    intent.needs_writable = true;
                    intent
                        .reasoning
                        .push(format!("Filesystem write needed: found `{indicator}`"));
                    break;
                }
            }
        }
    }

    // Extract specific paths
    extract_writable_paths(intent, code);
}

fn extract_writable_paths(intent: &mut CodeIntent, code: &str) {
    // Look for common writable paths in string literals
    let common_paths = ["/tmp", "/output", "/data", "/workspace", "/home", "/app"];
    for path in &common_paths {
        if code.contains(path)
            && intent.needs_writable
            && !intent.writable_paths.contains(&path.to_string())
        {
            intent.writable_paths.push(path.to_string());
        }
    }

    // If writable but no specific paths found, default to /tmp
    if intent.needs_writable && intent.writable_paths.is_empty() {
        intent.writable_paths.push("/tmp".to_string());
    }
}

fn analyze_resources(intent: &mut CodeIntent, code: &str, _language: &str) {
    // Data-heavy libraries suggest more memory
    let heavy_libs = [
        "pandas", "numpy", "scipy", "tensorflow", "torch", "polars",
        "xgboost", "lightgbm", "transformers",
    ];
    for lib in &heavy_libs {
        if code.contains(lib) {
            intent.memory_hint = Some("512m".to_string());
            intent
                .reasoning
                .push(format!("Memory hint 512m: data library `{lib}` detected"));
            break;
        }
    }
}

fn determine_provider(intent: &mut CodeIntent) {
    if intent.needs_network || intent.needs_packages || intent.needs_writable {
        intent.provider = ProviderHint::Docker;
        intent.confidence = 0.8;
    } else {
        // Pure compute — WASM is ideal
        intent.provider = ProviderHint::Wasm;
        intent.confidence = 0.7;
    }
    intent.reasoning.push(format!(
        "Provider: {:?} (confidence: {:.0}%)",
        intent.provider,
        intent.confidence * 100.0
    ));
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_pure_compute_selects_wasm() {
        let intent = analyze("print(2 + 2)", "python");
        assert_eq!(intent.provider, ProviderHint::Wasm);
        assert!(!intent.needs_network);
        assert!(!intent.needs_writable);
    }

    #[test]
    fn test_requests_detects_network() {
        let intent = analyze(
            "import requests\nresponses = requests.get('https://api.openai.com/v1/chat')",
            "python",
        );
        assert_eq!(intent.provider, ProviderHint::Docker);
        assert!(intent.needs_network);
        assert!(intent.network_hosts.contains(&"api.openai.com".to_string()));
    }

    #[test]
    fn test_pip_install_detects_packages() {
        let intent = analyze(
            "import subprocess\nsubprocess.run(['pip', 'install', 'pandas'])",
            "python",
        );
        assert!(intent.needs_packages || intent.needs_network);
        assert_eq!(intent.provider, ProviderHint::Docker);
    }

    #[test]
    fn test_pandas_import_detects_memory() {
        let intent = analyze(
            "import pandas as pd\ndf = pd.read_csv('data.csv')",
            "python",
        );
        assert_eq!(intent.memory_hint, Some("512m".to_string()));
        assert_eq!(intent.provider, ProviderHint::Docker);
    }

    #[test]
    fn test_file_write_detects_writable() {
        let intent = analyze(
            "with open('/tmp/output.txt', 'w') as f:\n    f.write('hello')",
            "python",
        );
        assert!(intent.needs_writable);
        assert!(intent.writable_paths.contains(&"/tmp".to_string()));
    }

    #[test]
    fn test_curl_bash() {
        let intent = analyze(
            "curl https://api.github.com/repos/foo/bar | jq '.name'",
            "bash",
        );
        assert!(intent.needs_network);
        assert!(intent.network_hosts.contains(&"api.github.com".to_string()));
        assert_eq!(intent.provider, ProviderHint::Docker);
    }

    #[test]
    fn test_node_fetch() {
        let intent = analyze("import fetch from 'node-fetch';\nconst res = await fetch('https://jsonplaceholder.typicode.com/todos/1');", "node");
        assert!(intent.needs_network);
        assert!(intent
            .network_hosts
            .contains(&"jsonplaceholder.typicode.com".to_string()));
    }

    #[test]
    fn test_host_extraction() {
        let host = extract_host_from_url("https://api.openai.com/v1/chat");
        assert_eq!(host, Some("api.openai.com".to_string()));
    }

    #[test]
    fn test_host_extraction_with_port() {
        // localhost has no dot, so it's filtered out (we only extract FQDN hosts)
        let host = extract_host_from_url("http://localhost:8080/api");
        assert_eq!(host, None);

        let host = extract_host_from_url("http://api.example.com:8080/api");
        assert_eq!(host, Some("api.example.com".to_string()));
    }
}
